ReactWindows/ReactNative.Shared/Views/Control/ReactControlManager.cs

```diff
@@ -0,0 +1,494 @@
+            {
+#if WINDOWS_UWP
+                var frame = Window.Current?.Content as Frame;
+                frame?.Focus(FocusState.Programmatic);
+#else
+                Keyboard.ClearFocus();
+#endif
+            }
+        }
+
+        /// <summary>
+        /// Called when view is detached from view hierarchy and allows for 
+        /// additional cleanup by the <see cref="IViewManager"/> subclass.
+        /// </summary>
+        /// <param name="reactContext">The React context.</param>
+        /// <param name="view">The view.</param>
+        /// <remarks>
+        /// Be sure to call this base class method to register for pointer 
+        /// entered and pointer exited events.
+        /// </remarks>
+        public override void OnDropViewInstance(ThemedReactContext reactContext, ReactControl view)
+        {
+            base.OnDropViewInstance(reactContext, view);
+            if (view.IsTabStop)
+            {
+                view.GotFocus -= OnGotFocus;
+                view.LostFocus -= OnLostFocus;
+                view.KeyDown -= OnKeyDown;
+                view.KeyUp -= OnKeyUp;
+            }
+        }
+
+        private void OnGotFocus(object sender, RoutedEventArgs e)
+        {
+            if (e.OriginalSource == sender)
+            {
+                var view = (ReactControl)sender;
+                view.GetReactContext()
+                    .GetNativeModule<UIManagerModule>()
+                    .EventDispatcher
+                    .DispatchEvent(new FocusEvent(view.GetTag()));
+            }
+        }
+
+        private void OnLostFocus(object sender, RoutedEventArgs e)
+        {
+            if (e.OriginalSource == sender)
+            {
+                var view = (ReactControl)sender;
+                view.GetReactContext()
+                    .GetNativeModule<UIManagerModule>()
+                    .EventDispatcher
+                    .DispatchEvent(new BlurEvent(view.GetTag()));
+            }
+        }
+
+#if WINDOWS_UWP
+        private void OnKeyDown(object sender, KeyRoutedEventArgs e)
+#else
+        private void OnKeyDown(object sender, KeyEventArgs e)
+#endif
+        {
+            if (sender == e.OriginalSource)
+            {
+                var view = (ReactControl)sender;
+                var keyCode = e.Key.GetKeyCode();
+                e.Handled = view.HandledKeyDownKeys?.Contains(keyCode) ?? false;
+                view.GetReactContext()
+                    .GetNativeModule<UIManagerModule>()
+                    .EventDispatcher
+                    .DispatchEvent(
+                        new KeyEvent(
+                            KeyEvent.KeyDownEventString,
+                            view.GetTag(),
+                            keyCode));
+            }
+        }
+
+#if WINDOWS_UWP
+        private void OnKeyUp(object sender, KeyRoutedEventArgs e)
+#else
+        private void OnKeyUp(object sender, KeyEventArgs e)
+#endif
+        {
+            if (sender == e.OriginalSource)
```

Comment:
This is not a good heuristic. Even though this is a bubbling event in JavaScript, there is no concept of `Handled` / `Unhandled`, so parent components will not get `onKey[Up|Down]` events that a JavaScript component does not want to handle.  One example of where this will probably be needed is in implementing tab and arrow key accessibility for ListView.  The individual list items will likely be focusable and may have responders for `Space` and `Enter`, but they will want the arrow key events to bubble to the `ListView` parent so they can change the focus programmatically (unless arrow keys work by default for tab stop controls in the same parent).

Reply:
Let's see if we can learn something about how to design this from touch events. I believe React Native only listens for touch events on the RootView. Do you understand the rationale behind this design?

Also, it seems like touch events would have a similar problem regarding the `Handled` property. How do touch events deal with `Handled` and would that solution work for keyboard events as well?

---
In reply to: [143015754](https://github.com/Microsoft/react-native-windows/pull/1411#discussion_r143015754) [](ancestors = 143015754)

Reply:
@rigdern I believe touch events are sufficiently different from the focus stack and keyboard events. When I first started thinking about how to implement this, I thought we might want to do something similar to the touch responder architecture in React.js and simulate the focus stack entirely in JS. This would be *a lot* of work, and ultimately will always be riddled with bugs where behaviors are inconsistent with what's expected in native platform.

Currently, we don't deal with the `Handled` property on touch events because the events need to bubble to the root view. The only components that handle touch events are built-in gesture responders like ScrollViewer, FlipView, etc., and that's all handled via direct manipulation (not the UI thread touch events).

Reply:
> I believe touch events are sufficiently different from the focus stack and keyboard events.

Can you elaborate on how they're sufficiently different?

> When I first started thinking about how to implement this, I thought we might want to do something similar to the touch responder architecture in React.js and simulate the focus stack entirely in JS. This would be a lot of work

Can you elaborate on why it's a lot of work? For focus, wouldn't we just tell React Native to dispatch the event into the JavaScript engine to the view with the appropriate tag? And then React Native would take care of the bubbling and capture phases.

> Currently, we don't deal with the Handled property on touch events because the events need to bubble to the root view.

Does that mean if we listen to events on the RootView, we don't have to worry about the `Handled` property? An exception I could see is if we wanted to mark `Handled` to `true` immediately on the control that triggered the event to prevent one of its native ancestors from handling the event in an undesirable way.